src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -0,0 +1,236 @@
+        /// <param name="IfRootObject">True if the object is a root object, i.e, it was passed by the client, otherwise false.</param>
+        /// <param name="objects">Objects of the top-level type we want to perform a deep update on.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool ifRootObject, params T[] objects)
+        {
+            foreach (object obj in objects)
+            {
+                EntityDescriptor rootEntityDescriptor = this.RequestInfo.EntityTracker.GetEntityDescriptor(obj);
+                
+                // If we do not find an EntityDescriptor in the entitytracker
+                // for any of the provided objects then we throw an exception. 
+                if (rootEntityDescriptor == null)
+                {
+                    throw Error.InvalidOperation(Strings.Context_EntityNotContained);
+                }
+
+                // If it is a top-level object,
+                // meaning that it was sent as top-level object by the user
+                // then we add it to the top-level objects. 
+                if (ifRootObject)
+                { 
+                    bulkUpdateGraph.AddTopLevelDescriptors(rootEntityDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor.DescriptorKind == DescriptorKind.Entity)
+                    {
+                        EntityDescriptor entityDescriptor = descriptor as EntityDescriptor;
+
+                        if (entityDescriptor.ParentEntity != null)
+                        {
+                            if (entityDescriptor.ParentEntity.Equals(rootEntityDescriptor.Entity))
+                            {
+                                // this check avoids cyclic operations. 
+                                if (!bulkUpdateGraph.CheckIfLinkAlreadyExists(entityDescriptor))
+                                {
+                                    bulkUpdateGraph.AddRelatedDescriptor(rootEntityDescriptor, entityDescriptor);
+                                    this.BuildDescriptorGraph(descriptors, false, entityDescriptor.Entity);
+                                } 
+                            }
+                        }      
+                    }
+                    else if (descriptor.DescriptorKind == DescriptorKind.Link)
+                    {
+                        //The link descriptor cannot be a parent to another link/entity descriptor
+                        LinkDescriptor link = descriptor as LinkDescriptor;
+
+                        if (link.Source.Equals(rootEntityDescriptor.Entity))
+                        {
+                            // this check avoid cyclic operations. 
+                            if (!bulkUpdateGraph.CheckIfLinkAlreadyExists(link))

```

Comment:
Consider merging these 2 if-statements into a single compound condition. It can help manage the excess code nesting in this method.